src/Tools/dotnet-monitor/Egress/S3/MultiPartUploadStream.cs

```diff
@@ -0,0 +1,147 @@
+using Amazon.S3;
+using Amazon.S3.Model;
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Tools.Monitor.Egress.S3;
+
+internal class MultiPartUploadStream : Stream
+{
+    private readonly byte[] _buffer;
+    private int _offset;
+    private readonly string _bucketName;
+    private readonly string _objectKey;
+    private readonly string _uploadId;
+    private readonly IAmazonS3 _client;
+    private readonly List<PartETag> _parts = new();
+    public List<PartETag> Parts => _parts.ToList();
+    public bool Disposed { get; private set; }
+    private int _position;
+    public const int MinimumSize = 5 * 1024 * 1024; // the minimum size of an upload part (except for the last part)
+
+    public MultiPartUploadStream(IAmazonS3 client, string bucketName, string objectKey, string uploadId, int bufferSize)
+    {
+        bufferSize = Math.Max(bufferSize, MinimumSize); // has to be at least the minimum
+        _buffer = ArrayPool<byte>.Shared.Rent(bufferSize);
+        _offset = 0;
+        _client = client;
+        _bucketName = bucketName;
+        _objectKey = objectKey;
+        _uploadId = uploadId;
+    }
+
+    public override async Task FlushAsync(CancellationToken cancellationToken)
+    {
+        if (Disposed)
+            throw new ObjectDisposedException(nameof(MultiPartUploadStream));
+        if (_offset == 0 || _offset < MinimumSize)
+            return;
+        await DoWriteAsync(cancellationToken);
+    }
+
+    public async Task FinalizeAsync(CancellationToken cancellationToken)
+    {
+        if (Disposed)
+            throw new ObjectDisposedException(nameof(MultiPartUploadStream));
+        if (_offset == 0)
+            return;
+        await DoWriteAsync(cancellationToken);
+    }
+
+    public override void Flush()
+    {
+        throw new NotSupportedException();
+    }
+
+    public override int Read(byte[] buffer, int offset, int count)
+    {
+        throw new NotSupportedException();
+    }
+
+    public override long Seek(long offset, SeekOrigin origin)
+    {
+        throw new NotSupportedException();
+    }
+
+    public override void SetLength(long value)
+    {
+        throw new NotSupportedException();
+    }
+
+    public override async Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+    {
+        if (Disposed)
+            throw new ObjectDisposedException(nameof(MultiPartUploadStream));
+
+        int BytesAvailableInBuffer() { return _buffer.Length - _offset; }
```

Comment:
`_offset` seems to never be modified except when writing a part (which sets it to `0`). It seems that it is always `0`.